### DISCUSSION TITLE
Getting image assets links using the API

### DISCUSSION THREAD

**Discussion by morismael** [solidusio/solidus#4383] [Troubleshooting] [answered]
Hello. Hope you're doing well. I want to thank you for this great product. It's really awesome. 

I am building an ecommerce website with its own frontend so basically using Solidus  API and Backend only. On the frontpage of the store website I wanna show the categories to allow customers to shop by category. I'm able to get the available categories (as Taxons) but the JSON response doesn't contain the icon of the taxon. Is there a way i could get this from the API request ? Otherwise how can I have this?  Also is there a plan to provide this feature in the near future? 

Thank you.

**Comment by kennyadsl** [comment, marked as answer]
Starting from Solidus 3.1 (https://github.com/solidusio/solidus/pull/4039), we can configure the JSON response attributes of API's endpoints ([including the ones related to taxons](https://github.com/solidusio/solidus/blob/f71a764581b620a14201e868c90391ae4c6bfa15/api/lib/spree/api_configuration.rb#L58-L61)) with a standard preference. 

In your `config/initializers/spree.rb`, you can add the following:

```ruby
Spree::Api::Config.configure do |config|
  config.taxon_attributes += [ :api_icon_url ]
end
```

And defining a new method on the taxon to retrieve the right size of the icon that you want to use:

```ruby
# app/overrides/your_store_name/spree/taxon/add_api_icon_url

module YourStoreName
  module Spree
    module Taxon
      module AddApuIconUrl
        def api_icon_url
          icon.url(:small) # :or :original
        end

        ::Spree::Taxon.prepend self
      end
    end
  end
end
```


which will be reflected in [the corresponding jbuilder template](https://github.com/solidusio/solidus/blob/f71a764581b620a14201e868c90391ae4c6bfa15/api/app/views/spree/api/taxons/_taxon.json.jbuilder#L4). Please note that, the specific partial linked is also used elsewhere and depending on the attributes you'll add, it could cause some performance degradations of other endopoints as well. 


⚠️ I didn't try this solution on a real application, so please let me know if this works or I missed anything.

**Reply by morismael** [reply]
Thank you for your swift answer.

I am trying to implement as suggested but getting this below error pointing to the line `icon.url(:original):`

```
ActionView::Template::Error (undefined method `analyzed?' for #<Spree::ActiveStorageAdapter::Attachment:0x00007fc3d80e49a0>):
     5: else
     6:   json.(taxonomy, *taxonomy_attributes)
     7:   json.root do
     8:     json.(taxonomy.root, *taxon_attributes)
     9:     json.taxons(taxonomy.root.children) do |taxon|
    10:       json.(taxon, *taxon_attributes)
    11:     end
```
I am using version `3.1.5` of Solidus

```
 solidus (3.1.5)
      solidus_api (= 3.1.5)
      solidus_backend (= 3.1.5)
      solidus_core (= 3.1.5)
      solidus_frontend (= 3.1.5)
      solidus_sample (= 3.1.5)
    solidus_api (3.1.5)
      jbuilder (~> 2.8)
      kaminari-activerecord (~> 1.1)
      responders
      solidus_core (= 3.1.5)
```
Any ideas?

**Reply by kennyadsl** [reply]
Can you try with just `icon.url` instead? Maybe I was wrong and `:original` is not a supported size.

**Reply by morismael** [reply]
Same error, with both variants or without none  :-)

**Reply by Azeem838** [reply]
You might have to try either `icon.rails_representation_url` or if it requires a path then pass `only_path: true`.

There's also `icon.service_url` that you can try as well.

You might have to include the Rails `url_helpers` if these methods are missing.

**Reply by elia** [reply]
@morismael can you please share the full stack trace of the error, that's really useful in understanding where the error is originated 🔍

**Reply by morismael** [reply]
Sorry gents. It's my mistake I think. I had removed the icons from the sample taxons. Let me try get them back before I test again.

**Comment by morismael** [comment]
Issue solved: it was the attachment issue. After attaching images to taxons it's working fine. Many thanks.

**Reply by kennyadsl** [reply]
Thanks, can you please mark the comment as the answer?

**Reply by morismael** [reply]
Done 👍🏾
